tiledb/sm/serialization/array_schema.cc

```diff
@@ -510,20 +510,41 @@ Status domain_to_capnp(
   return Status::Ok();
 }
 
-Status domain_from_capnp(
-    const capnp::Domain::Reader& domain_reader,
-    tdb_unique_ptr<Domain>* domain) {
+/* Deserialize a domain from a cap'n proto object. */
+shared_ptr<Domain> domain_from_capnp(
+    const capnp::Domain::Reader& domain_reader) {
+  Status st;
+
+  // Deserialize datatype
   Datatype datatype = Datatype::ANY;
-  RETURN_NOT_OK(datatype_enum(domain_reader.getType(), &datatype));
-  domain->reset(tdb_new(Domain));
+  st = datatype_enum(domain_reader.getType(), &datatype);
+  if (!st.ok()) {
+    throw std::runtime_error(
+        "[Deserialization::domain_from_capnp] " +
+        std::string(domain_reader.getType().cStr()) +
+        " is not a valid datatype identifer.");
+  }
 
+  // Validate the datatype
+  try {
+    ensure_datatype_is_valid(datatype);
+  } catch (const std::exception& e) {
+    std::throw_with_nested(
+        std::runtime_error("[Deserialization::domain_from_capnp] "));
+  }
+
+  // Deserialize dimensions
+  // Note: Security validation delegated to invoked API
   auto dimensions = domain_reader.getDimensions();
+  std::vector<shared_ptr<Dimension>> dims;
+  unsigned dim_num = 0;
   for (auto dimension : dimensions) {
-    auto dim = dimension_from_capnp(dimension);
-    RETURN_NOT_OK((*domain)->add_dimension(dim));
+    shared_ptr<Dimension> dim{dimension_from_capnp(dimension)};
+    dims.emplace_back(dim);
```

Comment:
You don't want a temporary variable here. It's an extra copy constructor that has no net effect but costs in performance.